src/main/kotlin/edu/cornell/cs/apl/viaduct/analysis/NameAnalysis.kt

```diff
@@ -201,25 +202,45 @@ class NameAnalysis(val tree: Tree<Node, ProgramNode>) {
 
     // TODO: readers for other types of [Name]s
 
-    /** Iterate up the AST from the given node, perforiming the effects of [f]. **/
-    private fun Node.iterateUp(tree: Tree<Node, ProgramNode>, f: (Node) -> Unit) {
-        f(this)
-        when (val parent = tree.parent(this)) {
-            null -> return
-            else -> parent.iterateUp(tree, f)
+    private val DeclarationNode.queries: Set<QueryNode> by collectedAttribute(tree) { node ->
+        if (node is QueryNode) {
+            listOf(declaration(node) to node)
+        } else {
+            listOf()
         }
     }
 
-    /** Calculate all of the loops a given node is contained in. **/
-    fun involvedLoops(node: Node): List<InfiniteLoopNode> {
-        val result = mutableListOf<InfiniteLoopNode>()
-        node.iterateUp(tree) {
-            if (it is InfiniteLoopNode) {
-                result.add(it)
+    fun queries(node: DeclarationNode): Set<QueryNode> = node.queries
+
+    private val DeclarationNode.updates: Set<UpdateNode> by collectedAttribute(tree) { node ->
+        if (node is UpdateNode) {
+            listOf(declaration(node) to node)
+        } else {
+            listOf()
+        }
+    }
+
+    fun updates(node: DeclarationNode): Set<UpdateNode> = node.updates
+
+    fun uses(node: DeclarationNode): Set<Node> =
+        queries(node).union(updates(node))
+
+    private val Node.involvedLoops: List<InfiniteLoopNode> by attribute {
+        val loopsAbove =
+            when (val parent = tree.parent(this)) {
+                null -> persistentListOf<InfiniteLoopNode>()
+                else -> parent.involvedLoops
             }
+        if (this is InfiniteLoopNode) {
```

Comment:
This makes a loop involved in itself. Not sure it matters, but if we care to fix, you would have to do this in the parent check.

```Kotlin
when (val parent = tree.parent(this)) {
    null -> persistentListOf<InfiniteLoopNode>()
    is LoopNode -> parent + parent.involvedLoops
    else -> parent.involvedLoops
}
```

Also, do we want inner to outer or outer to inner?